File: java/containsDuplicatesII.java
class Solution {
    public boolean containsNearbyDuplicate(int[] nums, int k) {
        //store the last k values in a hash set for constant time lookup
        Set<Integer> lastKValues = new HashSet<>();
        for (int i = 0; i < nums.length; i++) {
            //if the next value is contained in the last k values, return true
            if (lastKValues.contains(nums[i])) {
                return true;
            }
            //add the next value and, if necessary, remove the first element
            lastKValues.add(nums[i]);
            if (i >= k) {
                lastKValues.remove(nums[i - k]);
            }
        }
        return false;
    }
}

File: python/contains_duplicates_ii.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        #store last k values in a set for constant time lookup
        last_k_values = set()
        for i, num in enumerate(nums):
            #check if next value is in the last k values
            if num in last_k_values:
                return True
            #add the next value to the set
            last_k_values.add(num)
            #if necessary, remove the first value of the set
            if i >= k:
                last_k_values.remove(nums[i - k])
        return False